## BCIS5110 - Kiva Analysis

3 points each.

In [1]:
import pandas as pd
import numpy as np

In [2]:
# This code appears in every demonstration Notebook.
# By default, when you run each cell, only the last output of the codes will show.
# This code makes all outputs of a cell show.
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

### In this assignment, we start to work on Kiva datasets.
Assignment 8. Q1 - Q9

1. Read in the three tables as Pandas DataFrame

In [3]:
# It may take a while to load.
loans = pd.read_csv('./kiva_ds_csv/loans.csv')
lenders = pd.read_csv('./kiva_ds_csv/lenders.csv')
loans_lenders = pd.read_csv('./kiva_ds_csv/loans_lenders.csv')

2. Display a sample of each data frame. Show the attributes. How many observations? How many columns and what are they?

In [4]:
loans.head()


LOAN_ID   LOAN_NAME ORIGINAL_LANGUAGE  \
0  1441457   Rajabguli           Russian   
1  1717155  Rosa Maria           English   
2  1510286       Ruth            English   
3  1570048     Clarita           English   
4   343721      Leizel           English   

                                         DESCRIPTION  \
0  Рачабгули очень благодарна за вашу поддержку и...   
1  Rosa Maria is 54 years old and widowed. Rosa M...   
2  Hello Kiva community!<br /><br />             ...   
3  Clarita is 48 years old with four children. Wi...   
4  Leizel is 26 years old and married with one ch...   

                              DESCRIPTION_TRANSLATED  FUNDED_AMOUNT  \
0  Rajabguli is very grateful for the support and...          300.0   
1  Rosa Maria is 54 years old and widowed. Rosa M...          250.0   
2  Hello Kiva community!<br /><br />             ...          200.0   
3  Clarita is 48 years old with four children. Wi...          300.0   
4                                                NaN          375.0   

   LOAN_AMOUNT  STATUS   IMAGE_ID  VIDEO_ID  ... LENDER_TERM  \
0        300.0  funded  2728538.0       NaN  ...        14.0   
1        250.0  funded  3094251.0       NaN  ...        13.0   
2        200.0  funded  2824782.0       NaN  ...        14.0   
3        300.0  funded  2900215.0       NaN  ...         8.0   
4        375.0  funded   883415.0       NaN  ...        11.0   

  NUM_LENDERS_TOTAL NUM_JOURNAL_ENTRIES NUM_BULK_ENTRIES  \
0                 7                   5                3   
1                10                   2                1   
2                 7                   3                2   
3                 1                   2                1   
4                14                   2                1   

                                                TAGS BORROWER_NAMES  \
0  #Repeat Borrower, volunteer_like, volunteer_pi...      Rajabguli   
1          #Widowed, #Elderly, #Woman-Owned Business     Rosa Maria   
2                                                NaN          Ruth    
3                                                NaN        Clarita   
4                                                NaN         Leizel   

  BORROWER_GENDERS  BORROWER_PICTURED REPAYMENT_INTERVAL  DISTRIBUTION_MODEL  
0           female               true            monthly       field_partner  
1           female               true            monthly       field_partner  
2           female               true            monthly       field_partner  
3           female               true            monthly       field_partner  
4           female               true            monthly       field_partner  

[5 rows x 34 columns]

In [5]:
loans.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2087236 entries, 0 to 2087235
Data columns (total 34 columns):
 #   Column                           Dtype  
---  ------                           -----  
 0   LOAN_ID                          int64  
 1   LOAN_NAME                        object 
 2   ORIGINAL_LANGUAGE                object 
 3   DESCRIPTION                      object 
 4   DESCRIPTION_TRANSLATED           object 
 5   FUNDED_AMOUNT                    float64
 6   LOAN_AMOUNT                      float64
 7   STATUS                           object 
 8   IMAGE_ID                         float64
 9   VIDEO_ID                         float64
 10  ACTIVITY_NAME                    object 
 11  SECTOR_NAME                      object 
 12  LOAN_USE                         object 
 13  COUNTRY_CODE                     object 
 14  COUNTRY_NAME                     object 
 15  TOWN_NAME                        object 
 16  CURRENCY_POLICY                  object 
 17  CURRENCY

In [6]:
loans.describe(include='all')

LOAN_ID        LOAN_NAME ORIGINAL_LANGUAGE  \
count   2.087236e+06          2036998           2041345   
unique           NaN           462123                 8   
top              NaN  Anonymous Group           English   
freq             NaN             7082           1409566   
mean    1.063005e+06              NaN               NaN   
std     6.126119e+05              NaN               NaN   
min     8.400000e+01              NaN               NaN   
25%     5.341718e+05              NaN               NaN   
50%     1.061804e+06              NaN               NaN   
75%     1.592261e+06              NaN               NaN   
max     2.132739e+06              NaN               NaN   

                                              DESCRIPTION  \
count                                             2041310   
unique                                            2037086   
top     Hello Kiva Community! <br /><br />Meet Mary, w...   
freq                                                   17   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                                   DESCRIPTION_TRANSLATED  FUNDED_AMOUNT  \
count                                             1631924   2.087236e+06   
unique                                            1627852            NaN   
top     Hello Kiva Community! <br /><br />Meet Mary, w...            NaN   
freq                                                   17            NaN   
mean                                                  NaN   7.737961e+02   
std                                                   NaN   1.588132e+03   
min                                                   NaN   0.000000e+00   
25%                                                   NaN   2.500000e+02   
50%                                                   NaN   4.750000e+02   
75%                                                   NaN   9.000000e+02   
max                                                   NaN   5.000000e+05   

         LOAN_AMOUNT   STATUS      IMAGE_ID     VIDEO_ID  ...   LENDER_TERM  \
count   2.087236e+06  2087236  2.041345e+06  1178.000000  ...  2.087212e+06   
unique           NaN        4           NaN          NaN  ...           NaN   
top              NaN   funded           NaN          NaN  ...           NaN   
freq             NaN  1978361           NaN          NaN  ...           NaN   
mean    8.152003e+02      NaN  2.117308e+06  3141.939728  ...  1.312064e+01   
std     1.667433e+03      NaN  1.063870e+06  1238.038455  ...  7.266331e+00   
min     2.500000e+01      NaN  2.410000e+02     7.000000  ...  1.000000e+00   
25%     2.750000e+02      NaN  1.314524e+06  2603.750000  ...  8.000000e+00   
50%     5.000000e+02      NaN  2.186181e+06  3137.500000  ...  1.300000e+01   
75%     9.750000e+02      NaN  2.931119e+06  4105.750000  ...  1.400000e+01   
max     5.000000e+05      NaN  4.251491e+06  4985.000000  ...  1.950000e+02   

       NUM_LENDERS_TOTAL NUM_JOURNAL_ENTRIES NUM_BULK_ENTRIES           TAGS  \
count       2.087236e+06        2.087236e+06     2.087236e+06        1144179   
unique               NaN                 NaN              NaN         219437   
top                  NaN                 NaN              NaN  user_favorite   
freq                 NaN                 NaN              NaN          77681   
mean        2.132579e+01        1.540647e+00     1.147318e+00            NaN   
std         4.298174e+01        1.083405e+00     5.210305e-01            NaN   
min         0.000000e+00        1.000000e+00     1.000000e+00            NaN   
25%         7.000000e+00        1.000000e+00     1.000000e+00        

In [7]:
lenders.head()

PERMANENT_NAME DISPLAY_NAME  MAIN_PIC_ID       CITY    STATE COUNTRY_CODE  \
0   mary20651775         Mary          NaN        NaN      NaN          NaN   
1     debbie5637       Debbie          NaN        NaN      NaN          NaN   
2   alistair3809     Alistair          NaN  Pickering  Ontario           CA   
3    christy5044      Christy          NaN    Houston    Texas           US   
4      mandi1608        mandi     704871.0        NaN      NaN          NaN   

   MEMBER_SINCE PERSONAL_URL OCCUPATION LOAN_BECAUSE OTHER_INFO  \
0    1558031322          NaN        NaN          NaN        NaN   
1    1366375328          NaN        NaN          NaN        NaN   
2    1366377329          NaN        NaN          NaN        NaN   
3    1366380785          NaN        NaN          NaN        NaN   
4    1211248627          NaN        NaN          NaN        NaN   

   LOAN_PURCHASE_NUM INVITED_BY  NUM_INVITED  
0                  1        NaN            0  
1                  3        Tia            0  
2                  8        NaN            0  
3                  1        NaN            0  
4                 16      Becky            0

In [8]:
lenders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1561460 entries, 0 to 1561459
Data columns (total 14 columns):
 #   Column             Non-Null Count    Dtype  
---  ------             --------------    -----  
 0   PERMANENT_NAME     1561460 non-null  object 
 1   DISPLAY_NAME       1559864 non-null  object 
 2   MAIN_PIC_ID        516622 non-null   float64
 3   CITY               545203 non-null   object 
 4   STATE              509854 non-null   object 
 5   COUNTRY_CODE       667189 non-null   object 
 6   MEMBER_SINCE       1561460 non-null  int64  
 7   PERSONAL_URL       48738 non-null    object 
 8   OCCUPATION         337140 non-null   object 
 9   LOAN_BECAUSE       152092 non-null   object 
 10  OTHER_INFO         87785 non-null    object 
 11  LOAN_PURCHASE_NUM  1561460 non-null  int64  
 12  INVITED_BY         415651 non-null   object 
 13  NUM_INVITED        1561460 non-null  int64  
dtypes: float64(1), int64(3), object(10)
memory usage: 166.8+ MB


In [9]:
lenders.describe(include='all')

PERMANENT_NAME DISPLAY_NAME   MAIN_PIC_ID           CITY   STATE  \
count         1561460      1559864  5.166220e+05         545203  509854   
unique        1561460       214624           NaN          49548    8527   
top      mary20651775        David           NaN  San Francisco      CA   
freq                1        13336           NaN           9186   55509   
mean              NaN          NaN  1.210805e+06            NaN     NaN   
std               NaN          NaN  8.430773e+05            NaN     NaN   
min               NaN          NaN  1.554000e+03            NaN     NaN   
25%               NaN          NaN  5.637495e+05            NaN     NaN   
50%               NaN          NaN  1.052296e+06            NaN     NaN   
75%               NaN          NaN  1.737497e+06            NaN     NaN   
max               NaN          NaN  4.251567e+06            NaN     NaN   

       COUNTRY_CODE  MEMBER_SINCE      PERSONAL_URL OCCUPATION LOAN_BECAUSE  \
count        667189  1.561460e+06             48738     337140       152092   
unique          225           NaN             48082      80636       132328   
top              US           NaN  www.facebook.com    Student       I can.   
freq         436829           NaN                21      23861         2323   
mean            NaN  1.368739e+09               NaN        NaN          NaN   
std             NaN  1.180036e+08               NaN        NaN          NaN   
min             NaN  1.136106e+09               NaN        NaN          NaN   
25%             NaN  1.267996e+09               NaN        NaN          NaN   
50%             NaN  1.365415e+09               NaN        NaN          NaN   
75%             NaN  1.460599e+09               NaN        NaN          NaN   
max             NaN  1.616225e+09               NaN        NaN          NaN   

       OTHER_INFO  LOAN_PURCHASE_NUM INVITED_BY   NUM_INVITED  
count       87785       1.561460e+06     415651  1.561460e+06  
unique      85086                NaN      43711           NaN  
top                              NaN       John           NaN  
freq          127                NaN      22397           NaN  
mean          NaN       2.136488e+01        NaN  2.954619e-01  
std           NaN       2.724401e+02        NaN  2.039119e+01  
min           NaN       1.000000e+00        NaN  0.000000e+00  
25%           NaN       1.000000e+00        NaN  0.000000e+00  
50%           NaN       3.000000e+00        NaN  0.000000e+00  
75%           NaN       1.000000e+01        NaN  0.000000e+00  
max           NaN       1.674180e+05        NaN  2.492600e+04

In [10]:
loans_lenders.head()

LOAN_ID                                            LENDERS
0   448636  darryl4005, darryl4005, jim34605816, lauren137...
1   465841  darryl4005, andres66029178, andres66029178, da...
2   349829  christophe2368, christophe2368, eringailjensen...
3   267272  brian9150, maureen4671, rob7436, joseph7314, l...
4   317190  marc9298, timothy8041, lisa9240, donald8717, t...

In [11]:
loans_lenders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2034151 entries, 0 to 2034150
Data columns (total 2 columns):
 #   Column   Dtype 
---  ------   ----- 
 0   LOAN_ID  int64 
 1   LENDERS  object
dtypes: int64(1), object(1)
memory usage: 31.0+ MB


In [12]:
loans_lenders.describe(include='all')

LOAN_ID      LENDERS
count   2.034151e+06      2034151
unique           NaN      1971231
top              NaN  barbara5610
freq             NaN         3178
mean    1.054972e+06          NaN
std     6.155732e+05          NaN
min     8.400000e+01          NaN
25%     5.223755e+05          NaN
50%     1.041937e+06          NaN
75%     1.595558e+06          NaN
max     2.132731e+06          NaN

3. Check for missing values of columns of the loans dataframe. You can use sum() (instead of any()) to find out the number of missing values.

In [13]:
loans.isna().sum()

LOAN_ID                                  0
LOAN_NAME                            50238
ORIGINAL_LANGUAGE                    45891
DESCRIPTION                          45926
DESCRIPTION_TRANSLATED              455312
FUNDED_AMOUNT                            0
LOAN_AMOUNT                              0
STATUS                                   0
IMAGE_ID                             45891
VIDEO_ID                           2086058
ACTIVITY_NAME                            0
SECTOR_NAME                              0
LOAN_USE                             45914
COUNTRY_CODE                            34
COUNTRY_NAME                             0
TOWN_NAME                           176356
CURRENCY_POLICY                          0
CURRENCY_EXCHANGE_COVERAGE_RATE     473884
CURRENCY                                 0
PARTNER_ID                           21175
POSTED_TIME                              0
PLANNED_EXPIRATION_TIME             371834
DISBURSE_TIME                         4015
RAISED_TIME

In [14]:
loans.isna().sum()/len(loans)

LOAN_ID                            0.000000
LOAN_NAME                          0.024069
ORIGINAL_LANGUAGE                  0.021986
DESCRIPTION                        0.022003
DESCRIPTION_TRANSLATED             0.218141
FUNDED_AMOUNT                      0.000000
LOAN_AMOUNT                        0.000000
STATUS                             0.000000
IMAGE_ID                           0.021986
VIDEO_ID                           0.999436
ACTIVITY_NAME                      0.000000
SECTOR_NAME                        0.000000
LOAN_USE                           0.021998
COUNTRY_CODE                       0.000016
COUNTRY_NAME                       0.000000
TOWN_NAME                          0.084493
CURRENCY_POLICY                    0.000000
CURRENCY_EXCHANGE_COVERAGE_RATE    0.227039
CURRENCY                           0.000000
PARTNER_ID                         0.010145
POSTED_TIME                        0.000000
PLANNED_EXPIRATION_TIME            0.178147
DISBURSE_TIME                   

4. How many missing values for 'VIDEO_ID'? What's the percentage? What should we do? Do it.

In [15]:
## There are 2086058 missing  values in VIDEO_ID column. The percentga is 99.99 percent. We should remove the column
## as there are more than 90 percent null values.
loans.drop(columns=['VIDEO_ID'],inplace=True)

5. After examination of variable features, we intend to drop the following variables at this stage: LOAN_NAME, LOAN_USE, COUNTRY_CODE, TOWN_NAME, PARTNER_ID, PLANNED_EXPIRATION_TIME, TAGS, BORROWER_NAMES.

In [16]:
loans.drop(columns=["LOAN_NAME", "LOAN_USE", "COUNTRY_CODE", "TOWN_NAME", "PARTNER_ID", "PLANNED_EXPIRATION_TIME", "TAGS", "BORROWER_NAMES"],inplace=True)

6. Our purpose is to build a model that predicts whether a loan can get funded based on loan features. Check the dependent variable 'STATUS' in the loans dataframe. How many unique values?

In [17]:
loans['STATUS'].unique()

array(['funded', 'expired', 'refunded', 'fundRaising'], dtype=object)

7. How many observations for each category of 'STATUS'? (Which function we have learned can do that?)

In [18]:
loans['STATUS'].value_counts()

funded         1978361
expired          92416
refunded          9182
fundRaising       7277
Name: STATUS, dtype: int64

8. Which suggests the loan is funded? Which not? Which should we ignore?<br>
1) Delete the records with the category we can ignore. Refer to Boolean indexing in Numpy. We can apply the same operation to DataFrame to select rows we want.<br>

In [19]:
loans=loans[loans['STATUS']!='fundRaising']

8. 2) Recode the three categories to two categories (funded, notfunded). Consider .str.replace().<br>
The recoded variable will be our target variable.

In [20]:
loans['STATUS']=loans['STATUS'].str.replace('expired','notfunded').replace('refunded','notfunded')

In [21]:
loans['STATUS'].value_counts()

funded       1978361
notfunded     101598
Name: STATUS, dtype: int64

9. Examine variable 'IMAGE_ID'. What does the missing values for this variable mean?<br>
We can consider recode this variable to indicate whether there is image provided. Replace the missing value with 0 and change all IMAGE_ID to 1. 

In [24]:
## Missing value for this variable means there is no IMAGE_ID. 
loans['IMAGE_ID']=loans['IMAGE_ID'].fillna(0)

In [25]:
loans['IMAGE_ID']=loans['IMAGE_ID'].apply(lambda x: x if x==0 else 1)

In [26]:
loans['IMAGE_ID'].value_counts()

1.0    2034068
0.0      45891
Name: IMAGE_ID, dtype: int64

### Here starts Assignment 9.
Q10 - Q19

10. Convert variables 'POSTED_TIME', 'RAISED_TIME' and 'DISBURSE_TIME' to datetime data type.

In [28]:
loans['POSTED_TIME']=pd.to_datetime(loans['POSTED_TIME'])
loans['RAISED_TIME']=pd.to_datetime(loans['RAISED_TIME'])
loans['DISBURSE_TIME']=pd.to_datetime(loans['DISBURSE_TIME'])


11. Get the year of 'POSTED_TIME'. Assign it to a new variable.

In [29]:
loans['POSTED_TIME_YEAR']=loans['POSTED_TIME'].apply(lambda x:x.year)


12. Get the number of days of the difference between 'POSTED_TIME'and 'RAISED_TIME'. Assign the value to a variable. The difference (substraction) between two datetime is also a datetime object. You can get the number of days using .dt.days.

In [33]:
loans['difference']=(loans['POSTED_TIME']-loans['RAISED_TIME'])
loans['difference']=loans['difference'].apply(lambda x: x.days)

13. Create a new boolean variable 'predisburse': if the 'DISBURSE_TIME' is before 'POSTED_TIME', it equals to True, otherwise False.

In [36]:
loans['predisburse']=loans.apply(lambda x: True if x['DISBURSE_TIME']<x['POSTED_TIME'] else False , axis=1)

In [37]:
loans['predisburse'].value_counts()

True     1904374
False     175585
Name: predisburse, dtype: int64

14. Exame variable 'CURRENCY_EXCHANGE_COVERAGE_RATE'. How many unique values for this variable? How many missing?<br>
Before we rush to replace missing values with mode, we ask the question: is the missing value actually missing or is it something else? <br>
We then read the policies on Kiva about exchange coverage. For instance, a lender gives 25 in US dollars and it changes to 250 in another currency. But when the money is repaid at 250 in the other currency, it only exchanges to 20 US dollars due to the exchange rate fluctuating. There is loss for the lender.<br>
The CURRENCY_EXCHANGE_COVERAGE_RATE measures how much such loss will be covered by the field partner. Some would cover 10% or 20%. So, it is more likely that the missing values indicating the field partner will not share the loss. So, it's more likely missing means 0.<br>

Replace the missing values of this variable with 0.

In [38]:
loans['CURRENCY_EXCHANGE_COVERAGE_RATE'].unique()

array([0.1, 0.2, nan, 0. ])

In [39]:
loans['CURRENCY_EXCHANGE_COVERAGE_RATE'].isna().sum()

471302

In [40]:
loans['CURRENCY_EXCHANGE_COVERAGE_RATE']=loans['CURRENCY_EXCHANGE_COVERAGE_RATE'].fillna(0.0)

15. What is the total number of loans in each sector? What is the total amount of loans in each sector? 

In [45]:
loans['SECTOR_NAME'].value_counts()

Agriculture       538143
Food              458363
Retail            398944
Services          141503
Clothing          111665
Housing           106590
Personal Use       80246
Education          73147
Transportation     49518
Arts               41586
Health             29644
Construction       24728
Manufacturing      20681
Entertainment       2624
Wholesale           2577
Name: SECTOR_NAME, dtype: int64

In [47]:
loans.groupby('SECTOR_NAME')['LOAN_AMOUNT'].sum()

SECTOR_NAME
Agriculture       408144900.0
Arts               40308550.0
Clothing          116299925.0
Construction       24199750.0
Education          68736625.0
Entertainment       3866650.0
Food              377291725.0
Health             29562125.0
Housing            66769700.0
Manufacturing      17397875.0
Personal Use       32812425.0
Retail            316941425.0
Services          152961200.0
Transportation     36386675.0
Wholesale           3752550.0
Name: LOAN_AMOUNT, dtype: float64

16. What is the total number of loans in each year? 

In [48]:
loans['POSTED_TIME_YEAR'].value_counts()

2018    229633
2017    225476
2019    219005
2016    197236
2015    181835
2020    176490
2014    174232
2013    140035
2012    133847
2011    113009
2010     92902
2009     81629
2008     50697
2021     34419
2007     26314
2006      3200
Name: POSTED_TIME_YEAR, dtype: int64

17. Create a pivot table to show the average loan amount and total number of loans across sector, country and year.

In [49]:
loans.pivot_table(index = 'SECTOR_NAME', columns = 'POSTED_TIME_YEAR', aggfunc = {"LOAN_AMOUNT": "mean", 'LOAN_AMOUNT': 'count'}) 

LOAN_AMOUNT                                              \
POSTED_TIME_YEAR        2006    2007     2008     2009     2010     2011   
SECTOR_NAME                                                                
Agriculture            397.0  4086.0   9131.0  14357.0  19302.0  26140.0   
Arts                    68.0  1140.0   1497.0   1609.0   1918.0   2300.0   
Clothing               298.0  2539.0   4392.0   5626.0   6007.0   7563.0   
Construction           101.0   618.0   1504.0   1569.0   2056.0   1966.0   
Education                NaN    29.0     94.0    297.0    674.0   1040.0   
Entertainment            7.0    16.0     48.0    104.0    185.0    262.0   
Food                  1016.0  8121.0  14273.0  23850.0  22663.0  28350.0   
Health                  54.0   285.0    476.0    739.0    735.0    854.0   
Housing                 42.0   219.0    648.0   1920.0   2975.0   3835.0   
Manufacturing           42.0   490.0    686.0   1263.0   1369.0   1480.0   
Personal Use             NaN     NaN    185.0    335.0   1060.0   1552.0   
Retail                 695.0  5373.0  11423.0  20599.0  23359.0  25034.0   
Services               436.0  2601.0   4540.0   6524.0   6947.0   8365.0   
Transportation          44.0   661.0   1556.0   2650.0   3482.0   4061.0   
Wholesale                NaN   136.0    244.0    187.0    170.0    207.0   

                                                                        \
POSTED_TIME_YEAR     2012     2013     2014     2015     2016     2017   
SECTOR_NAME                                                              
Agriculture       30834.0  33745.0  44036.0  50161.0  52647.0  61272.0   
Arts               2452.0   2430.0   2610.0   2983.0   3909.0   5106.0   
Clothing           8665.0   8376.0   9904.0   8832.0   8957.0  10141.0   
Construction       1911.0   1968.0   1889.0   1727.0   1648.0   1932.0   
Education          2443.0   4935.0   8338.0   7647.0   9959.0  10245.0   
Entertainment       297.0    234.0    219.0    225.0    245.0    271.0   
Food              32786.0  33148.0  36539.0  37809.0  39723.0  44313.0   
Health              937.0   1275.0   1637.0   2262.0   3446.0   3896.0   
Housing            5277.0   5871.0  10916.0   8055.0   9234.0  11158.0   
Manufacturing      1644.0   1782.0   1847.0   1816.0   1674.0   1795.0   
Personal Use       1830.0   2611.0   5643.0   7878.0  13046.0  17344.0   
Retail            29796.0  28315.0  34123.0  35667.0  35135.0  37530.0   
Services          10218.0  10834.0  12068.0  12074.0  13171.0  15539.0   
Transportation     4534.0   4288.0   4297.0   4495.0   4273.0   4750.0   
Wholesale           223.0    223.0    166.0    204.0    169.0    184.0   

                                                      
POSTED_TIME_YEAR     2018     2019     2020     2021  
SECTOR_NAME                                           
Agriculture       63460.0  59184.0  56708.0  12683.0  
Arts               6034.0   3879.0   3018.0    633.0  
Clothing          11378.0  10187.0   7455.0   1345.0  
Construction       2328.0   1947.0   1342.0    222.0  
Education         10153.0  10244.0   5414.0   1635.0  
Entertainment       222.0    175.0    107.0      7.0  
Food              47208.0  42606.0  39100.0   6858.0  
Health             5378.0   4234.0   2835.0    601.0  
Housing           12549.0  23963.0   8756.0   1172.0  
Manufacturing      1959.0   1580.0   1074.0    180.0  
Personal Use       9362.0  11417.0   6583.0   1400.0  
Retail            38212.0  33724.0  33967.0   5992.0  
Services          16059.0  12395.0   8314.0   1418.0  
Transportation     5124.0   3336.0   1715.0    252.0  
Wholesale           207.0    134.0    102.0     21.0

18. Examine the variable 'BORROWER_GENDERS'. There is group borrowing, so we can see multiple borrowers. As research suggests, female borrowers may be riskier than male. It may affect a loan's chance to get funded. We will calculate the percentage of female borrowers and assign it to a new variable. You may write a function and use apply(). 

In [58]:
def perce(value):
    list_all=value.split(',')
    list_all_female=[x for x in list_all if x=='female']
    value_count=len(list_all_female)/len(list_all)
    return value_count
    

In [59]:
loans['BORROWER_GENDERS_FEMALE_PERCENTAGE']=loans['BORROWER_GENDERS'].apply(lambda x: perce(str(x)))

19. Answer one of the descriptive questions proposed in your initial project report.

In [60]:
# what percentage of borrower having language English

In [64]:
loans['ORIGINAL_LANGUAGE'].value_counts()/len(loans['ORIGINAL_LANGUAGE'])*100

English       67.511571
Spanish       21.912451
French         5.057359
Russian        2.612792
Portuguese     0.540299
Vietnamese     0.094473
Indonesian     0.048222
Arabic         0.016491
Name: ORIGINAL_LANGUAGE, dtype: float64

In [65]:
# 67 percent of borrower are haviong english language

# Preeti Chaudhry